src/foam/util/UIDSupport.java

```diff
@@ -0,0 +1,151 @@
+/**
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.util;
+
+public class UIDSupport {
+  /**
+   * Modulo constant for UID checksum
+   */
+  public final static int CHECKSUM_MOD = 997;
+
+  /**
+   * Singleton instance
+   */
+  private final static UIDSupport instance__ = new UIDSupport();
+
+  /**
+   * A hard coded array used as permutation sequence. It only supports
+   * permutation of a string less than 30 digits. The part of a string over
+   * 30 digits will not be involved in permutation.
+   */
+  private final static int[] PERMUTATION_SEQ = new int[] {
+    9,  3,  7,  9,  5,  6,  2, 8,  1,  4,
+    10, 11,  8, 12,  6, 14,  6, 5, 16,  3,
+    17,  2, 20, 18, 24, 17, 25, 3, 16, 12
+  };
+
+  private UIDSupport() {}
+
+  public static UIDSupport getInstance() {
+    return instance__;
+  }
+
+  /**
+   * Permutate id string according to the {@link #PERMUTATION_SEQ}.
+   *
+   * The checksum in the id string will be shifted to the beginning of the
+   * output string and its value is bumped up to avoid leading zeros since they
+   * could be lost during long integer uid conversion from and to hex string
+   * then mess up the length and permutation sequence when calculating its hash.
+   *
+   * @param idStr Source id string to permutate
+   * @return Generated unique ID
+   */
+  public String permutate(String idStr) {
+    var l = idStr.length() - 3;
+    var checksum = Integer.parseInt(idStr.substring(l), 16) + 256;
+    var id = new char[l];
+    idStr.getChars(0, l, id, 0);
+
+    for ( int i = 0 ; i < l ; i++ ) {
+      int newI = PERMUTATION_SEQ[i];
+      char c = id[newI];
+      id[newI] = id[i];
+      id[i] = c;
+    }
+    return toHexString(checksum, 3) + String.valueOf(id);
+  }
+
+  /**
+   * Undo/reverse permutate the generated unique id returned from
+   * {@code permutate(idStr) }.
+   *
+   * @param idStr Generated unique id
+   * @return Source id string
+   */
+  public String undoPermutate(String idStr) {
+    var checksum = Integer.parseInt(idStr.substring(0, 3), 16) - 256;
+    var id = idStr.substring(3).toCharArray();
+
+    for ( int i = id.length - 1 ; i >= 0; i-- ) {
+      int newI = PERMUTATION_SEQ[i];
+      char c = id[newI];
+      id[newI] = id[i];
+      id[i] = c;
+    }
+    return String.valueOf(id) + toHexString(checksum, 3);
+  }
+
+  /**
+   * Calculate the hash value of a generated long integer unique id.
+   *
+   * @param uid Generated unique id
+   * @return Hash of the unique id
+   */
+  public int hash(long uid) {
+    return hash(Long.toHexString(uid));
```

Comment:
So we convert to long to a String and then pass to Hash(String), which then parses it back to an int so that it can perform arithmetic on it. Is that correct?

Reply:
Yes, that's correct since the string uid is after permutation and we need to undo permutate before taking the mod.

Reply:
I see.